lib/semaphore.js

```diff
@@ -36,8 +36,8 @@ class Semaphore {
   }
 
   leave() {
-    this.counter++;
     if (this.queue.length === 0) {
+      this.counter++;
```

Comment:
Moving `this.counter++` here we will increase counter only if queue is empty. I do not understand a logic behind this

Reply:
1. We have concurrency = 3 and queue size = 4. We want to run 8 parallel tasks with Semaphore.
2. We enter first 3 times fine, now counter = 0, queue is empty, but other incoming tasks must go into queue.
3. We tried to enter 4 times, now 4 operations waiting on timeout for Semaphore to be free to enter from queue.
4. We do leave 1 time. 1 item from queue fill Semaphore on leave. counter still = 0. queue length = 3
5. We leave 3 times more, now counter still = 0. queue length = 0
6. We leave last 3 times. Now queue length =0. counter = concurrency (3).

If it is still unclear, I made another test with comments

Reply:
Old counter will increment over concurrency size, because it works with queue items as well. So every queue item will add 1 to counter. This is why we must not do anything with counter until queue is empty again

Reply:
You can still run tests without fix to see this behavior. After we don't use semaphore it will be not empty and counter will be more than concurrency, in case where we left after each enter.